server/sonar-server/src/main/java/org/sonar/server/source/SourceService.java

```diff
@@ -17,53 +17,114 @@
+  }
+
+  public Optional<Iterable<String>> getLinesAsHtml(DbSession dbSession, String fileUuid, int from, int toInclusive) {
+    return getLines(dbSession, fileUuid, from, toInclusive, lineToHtml());
   }
 
   /**
-   * Decorated lines of source file.
+   * Returns a single line from a source file. {@code Optional.absent()} is returned if the
+   * file or the line do not exist.
+   * @param line starts from 1
    */
-  public List<String> getLinesAsHtml(String fileUuid, @Nullable Integer fromParam, @Nullable Integer toParam) {
-    int from = (Integer) ObjectUtils.defaultIfNull(fromParam, 1);
-    int to = (Integer) ObjectUtils.defaultIfNull(toParam, Integer.MAX_VALUE);
-    List<String> lines = Lists.newArrayList();
-    for (SourceLineDoc lineDoc : sourceLineIndex.getLines(fileUuid, from, to)) {
-      lines.add(sourceDecorator.getDecoratedSourceAsHtml(lineDoc.source(), lineDoc.highlighting(), lineDoc.symbols()));
+  public Optional<DbFileSources.Line> getLine(DbSession dbSession, String fileUuid, int line) {
+    verifyLine(line);
+    FileSourceDto dto = dbClient.fileSourceDao().selectSourceByFileUuid(dbSession, fileUuid);
+    if (dto == null) {
+      return Optional.absent();
+    }
+    DbFileSources.Data data = dto.getSourceData();
+    return FluentIterable.from(data.getLinesList())
+      .filter(new IsGreaterOrEqualThanLine(line))
+      .first();
+  }
+
+  private <E> Optional<Iterable<E>> getLines(DbSession dbSession, String fileUuid, int from, int toInclusive, Function<DbFileSources.Line, E> function) {
+    verifyLine(from);
+    Preconditions.checkArgument(toInclusive >= from, String.format("Line number must greater than or equal to %d, got %d", from, toInclusive));
+    FileSourceDto dto = dbClient.fileSourceDao().selectSourceByFileUuid(dbSession, fileUuid);
+    if (dto == null) {
+      return Optional.absent();
+    }
+    DbFileSources.Data data = dto.getSourceData();
+    return Optional.of((Iterable<E>) FluentIterable.from(data.getLinesList())
+      .filter(new IsGreaterOrEqualThanLine(from))
+      .limit(toInclusive - from + 1)
+      .transform(function));
+  }
+
+  private void verifyLine(int line) {
+    Preconditions.checkArgument(line >= 1, String.format("Line number must start at 1, got %d", line));
+  }
+
+  private Function<DbFileSources.Line, String> lineToHtml() {
+    return new Function<DbFileSources.Line, String>() {
+      @Override
+      public String apply(@Nonnull DbFileSources.Line line) {
+        return htmlDecorator.getDecoratedSourceAsHtml(line.getSource(), line.getHighlighting(), line.getSymbols());
+      }
+    };
+  }
+
+  private enum LineToRaw implements Function<DbFileSources.Line, String> {
+    INSTANCE;
+    @Override
+    public String apply(@Nonnull DbFileSources.Line line) {
+      return line.getSource();
+    }
+
+  }
+
+  private class IsGreaterOrEqualThanLine implements Predicate<DbFileSources.Line> {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this a "static" inner class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2694)